File: implementation/python/alignment/needleman_wunsch.py
import sys
import numpy as np

MATCH = 2
MISMATCH = -1
GAP = -1
GAP_CHAR = "-"

# -----------------------------jjj-----------------------------//

SUBSTITUTION_MATRIX = {
    "A": {"A": MATCH, "C": MISMATCH, "G": MISMATCH, "T": MISMATCH},
    "C": {"A": MISMATCH, "C": MATCH, "G": MISMATCH, "T": MISMATCH},
    "G": {"A": MISMATCH, "C": MISMATCH, "G": MATCH, "T": MISMATCH},
    "T": {"A": MISMATCH, "C": MISMATCH, "G": MISMATCH, "T": MATCH},
}

# Set the constants that represent the three directions, which are used in the traceback matrix
UP = 1
LEFT = 2
DIAG = 3


def init_matrices(nrow, ncol):
    score_matrix = np.zeros((nrow, ncol), np.int32)
    traceback_matrix = np.zeros((nrow, ncol), np.int32)
    for i in range(1, nrow):
        score_matrix[i][0] = score_matrix[i - 1][0] + GAP
        traceback_matrix[i][0] = UP
    for j in range(1, ncol):
        score_matrix[0][j] = score_matrix[0][j - 1] + GAP
        traceback_matrix[0][j] = LEFT
    return score_matrix, traceback_matrix


def compute_score_and_traceback_matrices(s1, s2, substitution_matrix):
    nrow, ncol = len(s1) + 1, len(s2) + 1
    score_matrix, traceback_matrix = init_matrices(nrow, ncol)
    for i in range(1, nrow):
        for j in range(1, ncol):
            up = score_matrix[i - 1][j] + GAP
            left = score_matrix[i][j - 1] + GAP
            diag = (
                score_matrix[i - 1][j - 1] + substitution_matrix[s1[i - 1]][s2[j - 1]]
            )
            max_score, direction = compute_max_score_and_direction(up, left, diag)
            score_matrix[i][j] = max_score
            traceback_matrix[i][j] = direction
    return score_matrix, traceback_matrix


def traceback(traceback_matrix, s1, s2, gap_char):
    s1, s2 = list(s1), list(s2)
    i, j = len(s1), len(s2)
    aln1, aln2 = [], []
    while i > 0 and j > 0:
        direction = traceback_matrix[i][j]
        if direction == UP:
            aln1.append(s1.pop())
            aln2.append(gap_char)
            i -= 1
        elif direction == LEFT:
            aln1.append(gap_char)
            aln2.append(s2.pop())
            j -= 1
        else:
            aln1.append(s1.pop())
            aln2.append(s2.pop())
            i -= 1
            j -= 1
    return ("".join(reversed(aln1)), "".join(reversed(aln2)))


def needleman_wunsch(
    s1, s2, gap_char=GAP_CHAR, substitution_matrix=SUBSTITUTION_MATRIX
):
    score_matrix, traceback_matrix = compute_score_and_traceback_matrices(
        s1, s2, substitution_matrix
    )
    max_score = score_matrix[-1, -1]
    aln1, aln2 = traceback(traceback_matrix, s1, s2, gap_char)
    return (max_score, aln1, aln2)


def compute_max_score_and_direction(up, left, diag):
    max_score = max(up, left, diag)
    if max_score == up:
        direction = UP
    elif max_score == left:
        direction = LEFT
    else:
        direction = DIAG
    return max_score, direction


if __name__ == "__main__":
    s1, s2 = sys.argv[1:3]
    print(needleman_wunsch(s1, s2))

File: implementation/js/alignment/needleman_wunsch.js
import { show_alignment } from "./_show.js";
import { GAP, GAP_CHAR, SUBSTITUTION_MATRIX, UP, LEFT, DIAG } from "./_params.js";

// main
function main() {
  let [s1, s2] = process.argv.slice(2);
  let [s, a1, a2] = needleman_wunsch(s1, s2);
  console.log(show_alignment(a1, a2));
}

function needleman_wunsch(s1, s2, substitution_matrix = SUBSTITUTION_MATRIX, gap_penalty = GAP, gap_char = GAP_CHAR) {
  let [score_matrix, traceback_matrix] = compute_score_and_traceback_matrices(s1, s2, substitution_matrix, gap_penalty);
  let score = score_matrix[s1.length][s2.length];
  let [aln1, aln2] = traceback(traceback_matrix, s1, s2, gap_char);
  return [score, aln1, aln2];
}

function compute_score_and_traceback_matrices(s1, s2, substitution_matrix, gap_penalty) {
  let [nrow, ncol] = [s1.length + 1, s2.length + 1];
  let [score_matrix, traceback_matrix] = init_matrices([nrow, ncol], gap_penalty);
  for (let i = 1; i < nrow; i++) {
    for (let j = 1; j < ncol; j++) {
      let up = score_matrix[i - 1][j] + gap_penalty;
      let left = score_matrix[i][j - 1] + gap_penalty;
      let diag = score_matrix[i - 1][j - 1] + substitution_matrix[s1[i - 1]][s2[j - 1]];
      let [max_score, directions] = compute_max_score_and_direction(up, left, diag);
      score_matrix[i][j] = max_score;
      traceback_matrix[i][j] = directions;
    }
  }
  return [score_matrix, traceback_matrix];
}

// from the traceback matrix, resolve one of the possible alignments
function traceback(traceback_matrix, s1, s2, gap_char) {
  [s1, s2] = [Array.from(s1), Array.from(s2)];
  let [i, j] = [s1.length, s2.length];
  let [aln1, aln2] = [[], []];
  while (i > 0 && j > 0) {
    let direction = traceback_matrix[i][j];
    if (direction === UP) {
      aln1.push(s1.pop());
      aln2.push(gap_char);
      i--;
    } else if (direction === LEFT) {
      aln1.push(gap_char);
      aln2.push(s2.pop());
      j--;
    } else {
      aln1.push(s1.pop());
      aln2.push(s2.pop());
      i--;
      j--;
    }
  }
  return [aln1.reverse().join(""), aln2.reverse().join("")];
}

// initialise the score matrix and the traceback matrix
function init_matrices([nrow, ncol], gap_penalty) {
  let score_matrix = Array.from({ length: nrow }, (e) => Array(ncol).fill(0));
  let traceback_matrix = Array.from({ length: nrow }, (e) => Array(ncol).fill(0));
  for (let j = 1; j < ncol; j++) {
    score_matrix[0][j] = score_matrix[0][j - 1] + gap_penalty;
    traceback_matrix[0][j] = LEFT;
  }
  for (let i = 1; i < nrow; i++) {
    score_matrix[i][0] = score_matrix[i - 1][0] + gap_penalty;
    traceback_matrix[i][0] = UP;
  }
  return [score_matrix, traceback_matrix];
}

// compute the max score of one grid and one of the possible directions
function compute_max_score_and_direction(up, left, diag) {
  let max_score = Math.max(up, left, diag);
  let direction = up === max_score ? UP : left === max_score ? LEFT : DIAG;
  return [max_score, direction];
}

main();
